Sys 1009/marc record display improvements

kjallen opened the pull request:
PR for improved marc record display, some minor additions of css. To test display, submit form with existing marc record id, in the dev database, an existing sample is authority record id is `3228675`

The fields should be displayed in a tabular format, with columns of tag, indicators and data/subfields. The leader is presented as a header. 

Formatting was done via Black, a notable difference is double quotes rather than single in some cases. I can use other formatters for consistency if preferred. 

![Screen Shot 2022-10-05 at 4 29 42 PM](https://user-images.githubusercontent.com/1273423/194182677-43761424-c984-4a12-b859-7aadee93baf1.png)

kjallen committed:
Changed marc display to dictionary with attributes, in progress

kjallen committed:
Added css, table and proper display

kjallen committed:
Reformatting with Black

akohler commented on voyager_archive/templates/voyager_archive/marc_display.html:
```diff
@@ -11,18 +11,33 @@
-{% endcomment %}
+<div class="row">
+    <div class="column">
+        <h2>Record Leader: {{ marc_record.leader }}</h1>
```
I suggest dropping this to `h4` (and make start /end tags consistent :) ).  The leader is more critical than the rest of the MARC data.

akohler commented on voyager_archive/templates/voyager_archive/marc_display.html:
```diff
@@ -11,18 +11,33 @@
+    <div class="column">
+        <h2>Record Leader: {{ marc_record.leader }}</h1>
+        <table class="data-display">
+            <tr>
```
I recommend removing the table heading row.  The column headings are not important; users know the 3 parts, and they're in the conventional tag - indicators - field order.  This will tighten up the display, but I think it will still be at least as readable as it is now.

If column headings do prove more readable, then shorten `Indicators` to `Ind` and `Data and Subfields` to `Field`.

akohler commented on voyager_archive/static/css/style.css:
```diff
@@ -27,4 +27,12 @@
+    color: green
+  }
+  
+  .subfield {
```
Since this styles the subfield code aka the delimiter, and `delimiter` is the data used in the template, change this class to `delimiter`.

akohler commented on voyager_archive/templates/voyager_archive/marc_display.html:
```diff
@@ -11,18 +11,33 @@
+                <td>
+                    {{ field.data }}
+                    {% for delimiter, values in field.subfields.items %}
+                        <span class='subfield'>${{ delimiter }}</span>
```
Change the class to match the CSS update to `delimiter`.

akohler commented on voyager_archive/templates/voyager_archive/marc_display.html:
```diff
@@ -11,18 +11,33 @@
+                    {% for delimiter, values in field.subfields.items %}
+                        <span class='subfield'>${{ delimiter }}</span>
+                        {% for value in values %}
+                            {{ value }}
```
This needs to be reworked.  At present it displays only the values for each subfield - which is fine if a subfield occurs only once in a field, but wrong when a subfield occurs multiple times in a field.

Using bib 9411434 as an example, compare the 040 field.  It has a total of 7 subfields: `$a $b $e $c $d $d $d` (3 subfield $d).  But the current template shows this as `$a $b $e $c $d` (1 subfield $d), with all `$d` values in the same subfield.

![040_good](https://user-images.githubusercontent.com/2213836/194186309-6290160d-ca1e-4ae3-9b75-7ae893880d20.png)

![040_bad](https://user-images.githubusercontent.com/2213836/194186325-9a0aa48c-6684-43ca-89bb-efa323d522a6.png)

akohler commented on voyager_archive/templates/voyager_archive/marc_display.html:
```diff
@@ -11,18 +11,33 @@
+            {% for field in marc_record.fields %}
+            <tr>
+                <td>{{ field.tag }}</td>
+                <td><span class='indicators'>{{ field.indicators }}</span></td>
```
Blank indicators are not clear, at least when the first indicator is blank.  For example, in bib 9411434, the 050 field has 1st ind blank, 2nd ind 4.  The template currently displays that without the blank, so the 2nd ind 4 is aligned with the 1st ind of the 245 a couple of fields later.  See screenshots below.

You mentioned using a fixed width font, but I'm not seeing that defined in the CSS.  That could be one solution.

Other options:
* Change blank indicators to underscore for display
* Change blank indicators to `&nbsp;` (non-breaking space entity)
* Use separate table columns for each indicator after all

![indicators_bad](https://user-images.githubusercontent.com/2213836/194187323-2aa73555-724f-4b44-abee-66a39de0b821.png)

![indicators_good](https://user-images.githubusercontent.com/2213836/194187334-c632f138-7574-4e7e-99a1-5dcb094aede8.png)

akohler commented on voyager_archive/views_utils.py:
```diff
@@ -8,22 +8,55 @@ def get_auth_record(auth_id: int) -> list:
+
+        try:
+            subfield_dict = {}
+            subfield_default_dict = field.subfields_as_dict()
```
`subfields_as_dict()` combines multiple values into one list - which is OK at this point, but needs to be handled correctly for display, as noted in my comment above on bib 9411434 and its 040 field.

Confirming for myself :)
```
>>> f040.subfields_as_dict()
defaultdict(<class 'list'>, {'a': ['CLU'], 'b': ['eng'], 'e': ['rda'], 'c': ['CLU'], 'd': ['CLU', 'OCLCO', 'OCLCF']})
```

akohler reviewed (changes requested):
@kjallen - Excellent start.  There's one data problem with subfields, where repeated subfields must display individually with their delimiter (`$d val1 $d val2 $d val3`), not combined like `$d val1 val2 val3`.  That doesn't require any changes to the data structure, just include the delimiter with each subfield value as displayed.

There are a few other minor display changes as well.

Style formatting with `black` is fine, that's likely what I'll settle on once I decide to settle :)

Thanks --Andy

kjallen committed:
Change requests for formatting and indicator display

kjallen commented:
Corrections and additions:

- Corrected leader header to  h4 and consistent
- Changed subfield css class to delimiter
- Display delimiter in front of every entry for each subfield
- indicator1 and indicator2 can return various results if not present (empty string, single space, None...), strip white space and display underscore as default for missing/blank indicator for consistency

akohler commented on voyager_archive/views_utils.py:
```diff
@@ -8,22 +8,56 @@ def get_auth_record(auth_id: int) -> list:
     return get_marc_fields(marc_text)
 
+
 def get_marc_fields(marc_text: str) -> list:
```
Missed this before; the return type hint `list` should be changed to `dict`, since this method now returns a dictionary.

akohler commented on voyager_archive/views_utils.py:
```diff
@@ -8,22 +8,56 @@ def get_auth_record(auth_id: int) -> list:
+            tmp_rec_dict["data"] = field.data
+
+        if hasattr(field, "indicator1"):
+            tmp_rec_dict["indicator1"] = str.strip(field.indicator1)
```
Instead of removing blank indicators, then displaying underscores in the template, do that replacement here:
```
tmp_rec_dict["indicator1"] = field.indicator1.replace(" ", "_")
tmp_rec_dict["indicator2"] = field.indicator2.replace(" ", "_")
```

akohler commented on voyager_archive/templates/voyager_archive/marc_display.html:
```diff
@@ -11,18 +11,31 @@
+            <tr>
+                <td>{{ field.tag }}</td>
+                <td><span class='indicators'>
+                    {{ field.indicator1|default:"_" }}
```
Control fields (001-009) do not have indicators, so displaying `_ _` is not ideal.  With the blank -> underscore replacement done in `get_marc_fields()` (see other comment), the defaults are no longer needed here, so just
```
{{ field.indicator1 }}
{{ field.indicator2 }}
```

akohler reviewed (changes requested):
@kjallen The display is much improved, thanks.  A couple more minor tweaks to the indicators and we'll be done.

kjallen committed:
Corrected type hints and indicator substitution

kjallen commented:
PR updated to reflect corrections:
* type hints updated to reflect correct return type in view_utils
* indicator substitution updated to apply when appropriate

akohler reviewed (approved):
@kjallen Thanks for the fixes; merging --Andy
